support/scripts: Allow `mkcpio` to quietly gzip compress CPIO images

aadhi0319 opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://docs.unikraft.org/contribute.html

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://docs.unikraft.org/contribute.html) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->
Users may sometimes require compressed CPIO images. With the current
`mkcpio` script, the only way to do this is to create a regular CPIO
image and then gzip it separately. This is inefficient and unnecessary,
as `bsdcpio` natively allows for gzip compression. Furthermore, some
users may wish to suppress the output generated by the `mkcpio` script.
This commit offers new command line arguments that allow the user to
satisfy both of these requirements.

craciunoiuc commented:
Hey @aadhi0319,

Thanks for your contribution! Before accepting it I have a couple of questions:
 * You mention the "quiet" flag (`-q`), but it's not actually shown inside the "help" prompt. Is this intentional?
 * The default usage uses the verbose flag. Should this `-v` be added inside the "help" prompt?
 I'm thinking of the case where a user would assume that it's not on by default.

craciunoiuc reviewed (changes requested):
Hey @aadhi0319,

Thanks for your contribution! Before accepting it I have a couple of questions:
 * You mention the "quiet" flag (`-q`), but it's not actually shown inside the "help" prompt. Is this intentional?
 * The default usage uses the verbose flag. Should this `-v` be added inside the "help" prompt?
 I'm thinking of the case where a user would assume that it's not on by default.

aadhi0319 commented:
Hello @craciunoiuc!
* The quiet flag was a mistake on my part...I forgot to add it into the help prompt. I've added it in and you should see it with the next commit to this branch.
* I'm not sure I understand what you mean, so please correct me if I misunderstood. For bsdcpio, -v outputs the filenames that are being compressed. The number of blocks that have been written to the archive are also outputted regardless of the -v flag. --quiet suppresses writing the number of blocks written to the archive to stdout. For my use case, I wanted -q to output nothing (so -v was not set and --quiet was set). Not using -q sets -v and does not set --quiet, so the user will see the filenames being archived and the number of blocks written to the archive. Do you want me to split this into a separate -v and -q flags so that -v controls filenames being outputted and -q controls blocks written being outputted? Further, currently this script outputs the filenames and blocks written by default. As such, if we change the default behavior to be silent, this may require some programs that currently use this script to be altered.

craciunoiuc commented:
> Hello @craciunoiuc!
> 
>     * I'm not sure I understand what you mean, so please correct me if I misunderstood. For bsdcpio, -v outputs the filenames that are being compressed. The number of blocks that have been written to the archive are also outputted regardless of the -v flag. --quiet suppresses writing the number of blocks written to the archive to stdout. For my use case, I wanted -q to output nothing (so -v was not set and --quiet was set). Not using -q sets -v and does not set --quiet, so the user will see the filenames being archived and the number of blocks written to the archive. Do you want me to split this into a separate -v and -q flags so that -v controls filenames being outputted and -q controls blocks written being outputted? Further, currently this script outputs the filenames and blocks written by default. As such, if we change the default behavior to be silent, this may require some programs that currently use this script to be altered.

Hey,

Thanks for the verbose description :wink: . I was only curious about the default behaviour of `bsdcpio`.
There is no need to change anything.

I will add the `reviewed` as soon as you add the change :smiley:

craciunoiuc commented:
> Hello @craciunoiuc!
> 
>     * I'm not sure I understand what you mean, so please correct me if I misunderstood. For bsdcpio, -v outputs the filenames that are being compressed. The number of blocks that have been written to the archive are also outputted regardless of the -v flag. --quiet suppresses writing the number of blocks written to the archive to stdout. For my use case, I wanted -q to output nothing (so -v was not set and --quiet was set). Not using -q sets -v and does not set --quiet, so the user will see the filenames being archived and the number of blocks written to the archive. Do you want me to split this into a separate -v and -q flags so that -v controls filenames being outputted and -q controls blocks written being outputted? Further, currently this script outputs the filenames and blocks written by default. As such, if we change the default behavior to be silent, this may require some programs that currently use this script to be altered.

Hey,

Thanks for the verbose description :wink: . I was only curious about the default behaviour of `bsdcpio`.
There is no need to change anything.

craciunoiuc commented:
Hey @aadhi0319 @skuenzer,

The PR looks good now.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

skuenzer commented:
:thinking: The `quiet` option is fine - I would approve this. However, compressed (gzip) cpio archives aren't supported by Unikraft at the moment. I would delay that option until Unikraft supports it. Alternatively, we would need to document it (docs or tool help). I am wondering what you guys think because the intention of the script was to prepare a CPIO in the special format as Unikraft's `ukcpio` library currently needs...

skuenzer commented on support/scripts/mkcpio:
```diff
@@ -3,10 +3,37 @@ CPIOTOOL="$( which bsdcpio )"
-	echo "$0 [output.cpio] [CPIO root path]" 1>&2
+	echo "$0 [options] [output.cpio] [CPIO root path]" 1>&2
+	echo "-h	display this help screen" 1>&2
+	echo "-q	suppress non-critical output"
```
I guess a 1>&2 is missing here

aadhi0319 replied:
The changes to this script were made to facilitate the grub build script. I required a quiet mode and I thought a compressed cpio image would be better. As for now, I have not figured out how to get a compressed cpio image to work with multiboot anyways, so the compression option is never used. For that reason, I think it would be fine to delay the compressed option until later. As for a a default quiet behavior, I didn't want to break the script for anyone that assumed a list of files would be outputted by default. I wasn't sure who used the script or how they used it, but if it doesn't break anything I think quiet by default would be preferred to maintain consistency with other Unix commands.

skuenzer commented on support/scripts/mkcpio:
```diff
@@ -3,10 +3,37 @@ CPIOTOOL="$( which bsdcpio )"
 }
 
+# Process extra options
+FLAGS=""
```
From a style perspective, I would prefer building an Bash array:
```bash
CPIOFLAGS=()
```

skuenzer replied:
This is to get better control of argument splitting...

skuenzer commented on support/scripts/mkcpio:
```diff
@@ -3,10 +3,37 @@ CPIOTOOL="$( which bsdcpio )"
+			usage
+			;;
+		q)
+			FLAGS+="--quiet "
```
You add an argument then with
```bash
CPIOFLAGS+=("--quiet")
```

skuenzer commented on support/scripts/mkcpio:
```diff
@@ -3,10 +3,37 @@ CPIOTOOL="$( which bsdcpio )"
+			FLAGS+="--quiet "
+			;;
+		z)
+			FLAGS+="-z "
```
```bash
CPIOFLAGS+=("-z")
```

skuenzer commented on support/scripts/mkcpio:
```diff
@@ -29,8 +56,9 @@ if [ -z "${CPIOTOOL}" ]; then
 find . -depth \( ! -iname ".gitkeep" \) -print \
 	| tac \
-	| "${CPIOTOOL}" -v -o --format newc > "${OUT_BASE}/${OUT_FILE}"
+	| "${CPIOTOOL}" ${FLAGS} -o --format newc > "${OUT_BASE}/${OUT_FILE}"
```
To keep proper quoting of the individual arguments, you can then let bash expand the list variable with the `@` paramter:
```bash
"${CPIOFLAGS[@]}"
```

skuenzer commented on support/scripts/mkcpio:
```diff
@@ -3,10 +3,37 @@ CPIOTOOL="$( which bsdcpio )"
+			;;
+	esac
+done
+if [[ $FLAGS != *"--quiet"* ]]; then
```
Maybe it is more efficient to work with an boolean like variable that you set within the case:
```bash
OPT_QUIET=1
# [...]

while getopts ":hqz" OPT; do
	case $OPT in
		# [...]
		q)
			OPT_QUIET=0
			;;
		# [...]
	esac
done

if [ $OPT_QUIET -eq 0 ]; then
	CPIOFLAGS+=("--quiet")
else
	CPIOFLAGS+=("-v")	
fi
```

aadhi0319 replied:
I didn't know bash could expand arrays like that thanks for the tip! I agree with all the other style suggestions and will get a commit in with those fixes today or tomorrow.

skuenzer reviewed (commented):
Hi Aadhithya,
while checking the commit I found only one missing redirection to stderr (`1>&2`) on line 8. The other comments are just style suggestions from my side, nothing critical and not mandatory to change. I am also wondering if we should rather be in quiet mode as default (as like the default behavior for every Unix command) and provide a verbose option `-v`. But this is probably also just a matter of preference - what do you think?
Thanks,
Simon

craciunoiuc commented:
Hey @aadhi0319,

The PR looks good now. One mention: you have 2 commits with the same name that do different things. I suppose you wanted to 'squash' them together into a single one.

If yes, you can just:
```
git reset --soft HEAD~2
git commit -s -m "..."

craciunoiuc reviewed (changes requested):
Hey @aadhi0319,

The PR looks good now. One mention: you have 2 commits with the same name that do different things. I suppose you wanted to 'squash' them together into a single one.

If yes, you can just:
```
git reset --soft HEAD~2
git commit -s -m "..."
git push --force
```

aadhi0319 commented:
@craciunoiuc Thanks! Just squashed the commits.

craciunoiuc reviewed (approved):
All good now!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

skuenzer commented:
The PR looks fine so far. Probably you should update your commit message and title to be more to the point. How about something like:
```
support/scripts: mkcpio: Option for quietly creating archives

This commit introduces a `quiet` option to `mkcpio`. This option
suppresses standard output while creating an archive. It can be
useful when `mkcpio` is called from other scripts or Makefiles
(e.g., grub tools).
```

aadhi0319 committed:
support/scripts: mkcpio: Option for quietly creating archives

This commit introduces a `quiet` option to `mkcpio`. This option
suppresses standard output while creating an archive. It can be
useful when `mkcpio` is called from other scripts or Makefiles
(e.g., grub tools).

Signed-off-by: Aadhithya Kannan <aadhi0319@gmail.com>

aadhi0319 commented:
@skuenzer I pushed again with an updated commit message.

skuenzer reviewed (approved):
Thanks a lot for your work!

Approved-by: Simon Kuenzer <simon@unikraft.io>
